src/util/type.cpp

```diff
@@ -106,3 +106,52 @@ bool is_constant_or_has_constant_components(
 
   return false;
 }
+
+static std::string as_string(const struct_typet &src)
+{
+  std::string result="struct";
+  result+=" {";
+  bool first=true;
+
+  for(const auto &c : src.components())
+  {
+    if(first)
+      first=false;
+    else
+      result+=',';
+
+    result+=' ';
+
+    result+=c.type().as_string();
+    result+=' ';
+    result+=id2string(c.get_name());
+  }
+
+  result+=" }";
+
+  return result;
+}
+
+// The below generates a string in a generic syntax
+// that is inspired by C/C++/Java, and is meant for debugging
+// purposes.
+std::string typet::as_string() const
+{
+  if(id()==ID_pointer)
+    return '*'+subtype().as_string();
+  else if(id()==ID_array)
+  {
+    const auto &t=to_array_type(*this);
+    if(t.is_complete())
+      return subtype().as_string()+'['+t.size().as_string()+']';
+    else
+      return subtype().as_string()+"[]";
+  }
+  else if(id()==ID_struct)
+    return ::as_string(to_struct_type(*this));
+  else if(id()==ID_symbol)
+    return "symbol_type "+
+           id2string(to_symbol_type(*this).get_identifier());
+  else
+    return id_string();
```

Comment:
As with exprs, surely we want the named subs etc, especially for signedbv, floatbv etc.